File: src/converter/nameResolvers/resolveParenthesizedTypeName.ts
import ts from "typescript";
import {NameResolver} from "../nameResolver.js";
import {TypeScriptService, typeScriptServiceKey} from "../plugins/TypeScriptPlugin.js";

export function resolveParenthesizedTypeName(resolver: NameResolver): NameResolver {
    const parenthesizedResolver: NameResolver = (node, context) => {
        const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
        const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? ((node: ts.Node) => node.parent)

        const parent = getParent(node)
        if (
            !parent
            || !ts.isParenthesizedTypeNode(parent)
        ) {
            return resolver(node, context)
        }

        return parenthesizedResolver(parent, context)
    }

    return parenthesizedResolver
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/nameResolvers/resolveParenthesizedTypeName.kt
package io.github.sgrishchenko.karakum.extension.nameResolvers

import io.github.sgrishchenko.karakum.extension.ConverterContext
import io.github.sgrishchenko.karakum.extension.NameResolver
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.Node
import typescript.isParenthesizedTypeNode

fun resolveParenthesizedTypeName(resolver: NameResolver<Node>): NameResolver<Node> {
    fun parenthesizedResolver(node: Node, context: ConverterContext): String? {
        val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)
        val getParent = { it: Node ->
            typeScriptService?.getParent(it) ?: it.getParentOrNull()
        }

        val parent = getParent(node)
        if (
            parent == null
            || !isParenthesizedTypeNode(parent)
        ) {
            return resolver(node, context)
        }

        return parenthesizedResolver(parent, context)
    }

    return NameResolver { node, context -> parenthesizedResolver(node, context) }
}
